File: library/src/main/java/com/vorlonsoft/android/rate/Time.java
/*
 * Copyright 2018 Vorlonsoft LLC
 *
 * Licensed under The MIT License (MIT)
 */

package com.vorlonsoft.android.rate;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;

import androidx.annotation.LongDef;

import static com.vorlonsoft.android.rate.Constants.Utils.UTILITY_CLASS_MESSAGE;
import static java.lang.annotation.RetentionPolicy.SOURCE;

/**
 * <p>Time Class - time units class of the AndroidRate library.</p>
 *
 * @since    1.2.1
 * @version  1.2.1
 * @author   Alexander Savin
 */

@SuppressWarnings("WeakerAccess")
public class Time {
    /** <p>Constant for unit of duration in milliseconds: millisecond.</p> */
    public static final long MILLISECOND = 1;
    /** <p>Constant for unit of duration in milliseconds: second.</p> */
    public static final long SECOND = MILLISECOND * 1000;
    /** <p>Constant for unit of duration in milliseconds: minute.</p> */
    public static final long MINUTE = SECOND * 60;
    /** <p>Constant for unit of duration in milliseconds: hour.</p> */
    public static final long HOUR = MINUTE * 60;
    /** <p>Constant for unit of duration in milliseconds: day.</p> */
    public static final long DAY = HOUR * 24;
    /** <p>Constant for unit of duration in milliseconds: week.</p> */
    public static final long WEEK = DAY * 7;
    /** <p>Constant for unit of duration in milliseconds: month.</p> */
    public static final long MONTH = DAY * 30;
    /** <p>Constant for unit of duration in milliseconds: year.</p> */
    public static final long YEAR = DAY * 365;

    private Time() {
        throw new UnsupportedOperationException("Time" + UTILITY_CLASS_MESSAGE);
    }

    /**
     * <p>Annotates element of long type.</p>
     * <p>Annotated element, represents a logical type and its value should be one of the following
     * constants: MILLISECOND, SECOND, MINUTE, HOUR, DAY, WEEK, MONTH, YEAR.</p>
     */
    @Documented
    @Retention(SOURCE)
    @LongDef({
            MILLISECOND,
            SECOND,
            MINUTE,
            HOUR,
            DAY,
            WEEK,
            MONTH,
            YEAR
    })
    public @interface TimeUnits {
    }
}

File: library/src/main/java/com/vorlonsoft/android/rate/Time.kt
/*
 * Copyright 2018 Vorlonsoft LLC
 *
 * Licensed under The MIT License (MIT)
 */

package com.vorlonsoft.android.rate

import androidx.annotation.LongDef

/**
 * Time Class - the time units class of the AndroidRate library.
 *
 * Contains time units constants and [TimeUnits] annotation.
 *
 * @constructor Don't create an instance of this class. Use its members directly.
 * @since       1.2.1
 * @version     1.2.1
 * @author      Alexander Savin
 */
open class Time {
    /**
     * Denotes that the annotated element of the primitive type long represents a logical type and
     * that its value should be one of the following constants: [MILLISECOND], [SECOND], [MINUTE],
     * [HOUR], [DAY], [WEEK], [MONTH], [YEAR].
     *
     * @since    1.2.1
     * @version  1.2.1
     * @author   Alexander Savin
     */
    @MustBeDocumented
    @Retention(AnnotationRetention.SOURCE)
    @LongDef(MILLISECOND, SECOND, MINUTE, HOUR, DAY, WEEK, MONTH, YEAR)
    annotation class TimeUnits

    /** Contains constants for units of time in milliseconds. */
    companion object {
        /** Time unit representing one thousandth of a second. */
        const val MILLISECOND: Long = 1L
        /** Time unit representing one second. */
        const val SECOND: Long = MILLISECOND * 1_000L
        /** Time unit representing sixty seconds. */
        const val MINUTE: Long = SECOND * 60L
        /** Time unit representing sixty minutes. */
        const val HOUR: Long = MINUTE * 60L
        /** Time unit representing twenty four hours. */
        const val DAY: Long = HOUR * 24L
        /** Time unit representing seven days. */
        const val WEEK: Long = DAY * 7L
        /** Time unit representing thirty days. */
        const val MONTH: Long = DAY * 30L
        /** Time unit representing three hundred and sixty five days. */
        const val YEAR: Long = DAY * 365L
    }
}